### PR TITLE
Implementation of RFC5077 SessionTickets resumption.

### LINKED ISSUE #60: Session tickets support [TLS 1.2]
Add support for [RFC 4507](http://tools.ietf.org/html/rfc4507)

### PR DESCRIPTION
**Pull request by inikolcev**
This enables both client and server side to resume session using session tickets defined in RFC5077.
I have tested this against openssl and gnutls, both client and server side in tls1.2, tls1.1 and tls1.0 and it seems to be working.
To enable it on client side all you need to is initialize an empty list for connection.session_ticket_cache where the tickets received from server are held.
Example:
```
session_cache = []
connection.session_ticket_cache = session_cache
```
To enable it on server side you need to generate the encryption keys that are held in settings.session_ticket_keys.
Example:
```
 settings.session_ticket_keys = {}
 settings.session_ticket_keys["key_name"] = getRandomBytes(16)
 settings.session_ticket_keys["aes_key"] = getRandomBytes(16)
 settings.session_ticket_keys["hmac_key"] = getRandomBytes(32)
```
This fixes #60

---
This change is [Reviewable](https://reviewable.io/reviews/tlsfuzzer/tlslite-ng/437)

### CODE REVIEW

**Review by tomato42 — CHANGES_REQUESTED** [review]
> To enable it on client side all you need to is initialize an empty list for connection.session_ticket_cache where the tickets received from server are held.

why we don't want them enabled by default? the TLS 1.3 tickets aren't enabled by default?

> To enable it on client side all you need to is initialize an empty list for connection.session_ticket_cache where the tickets received from server are held.

why do we want to burden the user with such detailed configuration? why not reuse the configuration from TLS 1.3 tickets, the `HandshakeSetting`'s `ticketKeys` and `ticketCipher`?

> Reviewed 7 of 7 files at r1, 3 of 3 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/tlsfuzzer/tlslite-ng/437#-:-MNcaIIcC-ptHzMtxuGn:b1153zk)* status: all files reviewed, 23 unresolved discussions (waiting on @inikolcev)

___
*[scripts/tls.py, line 340 at r3](https://reviewable.io/reviews/tlsfuzzer/tlslite-ng/437#-MNciVpaA6xpl6nzhO8N:-MNciVpaA6xpl6nzhO8O:bavbyr8) ([raw file](https://github.com/tlsfuzzer/tlslite-ng/blob/35b886b54cf94389ea60eb5be5b5e8681356fa50/scripts/tls.py#L340)):*
> ```Python
>         print("  Extended Master Secret: {0}".format(
>                                                    connection.extendedMasterSecret))
>         print("  Session Resumed: True")
> ```

that's a lot of code duplication...

and we really don't want the connection object to behave any different with and without resumption...
___
*[scripts/tls.py, line 424 at r3](https://reviewable.io/reviews/tlsfuzzer/tlslite-ng/437#-MNciiaS70PK3EWIA0dw:-MNciiaTDzDcLiCWwV3f:b-x9b47t) ([raw file](https://github.com/tlsfuzzer/tlslite-ng/blob/35b886b54cf94389ea60eb5be5b5e8681356fa50/scripts/tls.py#L424)):*
> ```Python
>     connection = TLSConnection(sock)
>     session_cache = []
>     connection.session_ticket_cache = session_cache
> ```

this shouldn't be necessary
___
*[tests/tlstest.py, line 1349 at r2](https://reviewable.io/reviews/tlsfuzzer/tlslite-ng/437#-MNchaGbAxlvgpBlyHsa:-MNchaGbAxlvgpBlyHsb:b-6rfem3) ([raw file](https://github.com/tlsfuzzer/tlslite-ng/blob/6ea7428daa7bfb0a34cb69d7bbf7fbdd0aa60335/tests/tlstest.py#L1349)):*
> ```Python
>     test_no += 1
> 
>     print("Test {0} - session_ticket resumption in TLSv1.2".format(test_no))
> ```

1. we should also check that the handshake will be successful even if the ticket we send is garbage
2. we should also have test coverage for client certificate authentication and verify that client cert is available to server post-resume
___
*[tests/tlstest.py, line 1369 at r2](https://reviewable.io/reviews/tlsfuzzer/tlslite-ng/437#-MNchOwQ5AVOwhzYgH5v:-MNchOwQ5AVOwhzYgH5w:b-h5upgc) ([raw file](https://github.com/tlsfuzzer/tlslite-ng/blob/6ea7428daa7bfb0a34cb69d7bbf7fbdd0aa60335/tests/tlstest.py#L1369)):*
> ```Python
>     connection.handshakeClientCert(serverName=address[0], settings=settings)
>     testConnClient(connection)
>     assert connection.resumed
> ```

assert that serverCertChain is still set missing
___
*[tlslite/extensions.py, line 2121 at r1](https://reviewable.io/reviews/tlsfuzzer/tlslite-ng/437#-MNcc_muBYOrR5SGsMlO:-MNcc_muBYOrR5SGsMlP:bfyedt8) ([raw file](https://github.com/tlsfuzzer/tlslite-ng/blob/c671f3dee381b05c3a8000e5fdfb8122cfbed7d6/tlslite/extensions.py#L2121)):*
> ```Python
>     def extData(self):
>         """Serialise the payload of the extension."""
>         if len(self.ticket) == 0:
> ```

`if not self.ticket:` is more pythonic
___
*[tlslite/handshakesettings.py, line 296 at r1](https://reviewable.io/reviews/tlsfuzzer/tlslite-ng/437#-MNccvWB8C06gIzn6qpE:-MNccvWB8C06gIzn6qpF:b-hom8i7) ([raw file](https://github.com/tlsfuzzer/tlslite-ng/blob/c671f3dee381b05c3a8000e5fdfb8122cfbed7d6/tlslite/handshakesettings.py#L296)):*
> ```Python
>         Leave empty to disable session ticket support on server side.
> 
>     :vartype session_ticket_keys: dict
> ```

as I wrote in the opening, why not reuse ticketKeys?
___
*[tlslite/messages.py, line 2058 at r1](https://reviewable.io/reviews/tlsfuzzer/tlslite-ng/437#-MNcdIGC1cPSpTx72FYS:-MNcdIGC1cPSpTx72FYT:b-3y5bir) ([raw file](https://github.com/tlsfuzzer/tlslite-ng/blob/c671f3dee381b05c3a8000e5fdfb8122cfbed7d6/tlslite/messages.py#L2058)):*
> ```Python
> 
> 
> class NewSessionTicket2(HelloMessage):
> ```

"2" we used to indicate SSLv2 objects.... I think "1_0" for TLS v1.0 (first version to support them) would be more appropriate
___
*[tlslite/messages.py, line 2170 at r1](https://reviewable.io/reviews/tlsfuzzer/tlslite-ng/437#-MNceB3U5cdKE9uVB2kQ:-MNceB3U5cdKE9uVB2kR:b-tnmx25) ([raw file](https://github.com/tlsfuzzer/tlslite-ng/blob/c671f3dee381b05c3a8000e5fdfb8122cfbed7d6/tlslite/messages.py#L2170)):*
> ```Python
>         self.protocol_version = (parser.get(1), parser.get(1))
>         self.cipher_suite = parser.get(2)
>         if expect_nonce:
> ```

we don't need a nonce??
___
*[tlslite/session.py, line 188 at r1](https://reviewable.io/reviews/tlsfuzzer/tlslite-ng/437#-MNceTI87opGOZLn4VQk:-MNceTI87opGOZLn4VQl:b-isdd3i) ([raw file](https://github.com/tlsfuzzer/tlslite-ng/blob/c671f3dee381b05c3a8000e5fdfb8122cfbed7d6/tlslite/session.py#L188)):*
> ```Python
>     """
>     This class holds the ticket and ticket lifetime which are recieved from the server,
>     together with master secret and ciphersuite which are needed to resume a session
> ```

no, that's far too little information to correctly resume a session in all cases, you need the complete session object and the Ticket should be just a field in it
___
*[tlslite/session.py, line 218 at r1](https://reviewable.io/reviews/tlsfuzzer/tlslite-ng/437#-MNcesMy3viIhzOpcOWI:-MNcesMy3viIhzOpcOWJ:bcp8fhi) ([raw file](https://github.com/tlsfuzzer/tlslite-ng/blob/c671f3dee381b05c3a8000e5fdfb8122cfbed7d6/tlslite/session.py#L218)):*
> ```Python
>     def valid(self):
>         if time.time() > self.time_received:
>             return True
> ```

invalid tickets should get automatically purged from cache
___
*[tlslite/tlsconnection.py, line 1579 at r1](https://reviewable.io/reviews/tlsfuzzer/tlslite-ng/437#-MNcf2Rv6IXdCeO5lKqJ:-MNcf2Rv6IXdCeO5lKqK:b-ppjr0b) ([raw file](https://github.com/tlsfuzzer/tlslite-ng/blob/c671f3dee381b05c3a8000e5fdfb8122cfbed7d6/tlslite/tlsconnection.py#L1579)):*
> ```Python
>                       nextProto=None, settings=None):
> 
>         # If we have SessionTickets enabled in TLS1.2, try to get a valid one
> ```

like I wrote above, no, to perform correct resumption in all cases you need the complete session object
___
*[tlslite/tlsconnection.py, line 2520 at r1](https://reviewable.io/reviews/tlsfuzzer/tlslite-ng/437#-MNcfaPu9WBDvLBePrJw:-MNcfaPu9WBDvLBePrJx:b-2u4234) ([raw file](https://github.com/tlsfuzzer/tlslite-ng/blob/c671f3dee381b05c3a8000e5fdfb8122cfbed7d6/tlslite/tlsconnection.py#L2520)):*
> ```Python
>             new_ticket = NewSessionTicket2()
>             new_ticket.create(settings.ticketLifetime,
>                               key_name + iv + hmac +
> ```

this is just duplicating the TLS 1.3 code...
___
*[tlslite/tlsconnection.py, line 4339 at r1](https://reviewable.io/reviews/tlsfuzzer/tlslite-ng/437#-MNcg54j8ubZ5mlcsSvV:-MNcg54j8ubZ5mlcsSvW:b3intpt) ([raw file](https://github.com/tlsfuzzer/tlslite-ng/blob/c671f3dee381b05c3a8000e5fdfb8122cfbed7d6/tlslite/tlsconnection.py#L4339)):*
> ```
> key_name
> ```
unused variable?
___
*[tlslite/tlsconnection.py, line 4377 at r1](https://reviewable.io/reviews/tlsfuzzer/tlslite-ng/437#-MNcgFpa3sreKA2TDtta:-MNcgFpa3sreKA2TDttb:b-hesohy) ([raw file](https://github.com/tlsfuzzer/tlslite-ng/blob/c671f3dee381b05c3a8000e5fdfb8122cfbed7d6/tlslite/tlsconnection.py#L4377)):*
> ```Python
>             settings=None, send_session_ticket=False, client_cert_chain=None):
> 
>         if send_session_ticket:
> ```

I wonder if this shouldn't be a field in the object rather than variable explicitly passed around
___
*[tlslite/tlsconnection.py, line 4449 at r1](https://reviewable.io/reviews/tlsfuzzer/tlslite-ng/437#-MNcgvle5PX7MlURNf3U:-MNcgvle5PX7MlURNf3V:b-qd5wvl) ([raw file](https://github.com/tlsfuzzer/tlslite-ng/blob/c671f3dee381b05c3a8000e5fdfb8122cfbed7d6/tlslite/tlsconnection.py#L4449)):*
> ```Python
> 
>                 changeCipherSpec = result
>                 if changeCipherSpec.type != 1:
> ```

duplicated code

**Comment by tomato42** [comment]
just a reminder: you'll need to rebase on top of current master, with https://github.com/tlsfuzzer/tlslite-ng/pull/467 merged, otherwise the py2.6 runs will fail.

**Review by tomato42 — COMMENTED** [review]
> Reviewed 3 of 8 files at r4.
> *[Reviewable](https://reviewable.io/reviews/tlsfuzzer/tlslite-ng/437)* status: 6 of 11 files reviewed, 21 unresolved discussions (waiting on @inikolcev and @tomato42)

**Review by tomato42 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tlsfuzzer/tlslite-ng/437)* status: 6 of 11 files reviewed, 22 unresolved discussions (waiting on @inikolcev and @tomato42)

___
*[tlslite/extensions.py, line 2111 at r4](https://reviewable.io/reviews/tlsfuzzer/tlslite-ng/437#-MZE_m9f1-pbLew4NhWD:-MZE_m9g2QNaJBmEE4dz:bvdsdnn) ([raw file](https://github.com/tlsfuzzer/tlslite-ng/blob/c464c42242f66739176a5d6303b6194380eaf847/tlslite/extensions.py#L2111)):*
> ```Python
>                                                      session_ticket)
>         self.ticket = None
>         self.ticket_length = None
> ```

`ticket_length` is unused, isn't it?

**Comment by tomato42** [comment]
looks like `tls.py server` sends two NewSessionTicket messages by default; that's not allowed in TLS 1.2...

**Review by tomato42 — COMMENTED** [review]
> Reviewed 4 of 8 files at r4.
> Dismissed @codeclimate[bot] from 5 discussions.
> *[Reviewable](https://reviewable.io/reviews/tlsfuzzer/tlslite-ng/437)* status: 10 of 11 files reviewed, 7 unresolved discussions (waiting on @inikolcev)

___
*[`tlslite/extensions.py` line 2111 at r4](https://reviewable.io/reviews/tlsfuzzer/tlslite-ng/437#-MZE_m9f1-pbLew4NhWD:-NNm7uvV4soM5Zx7N0_l:b-7ql1td) ([raw file](https://github.com/tlsfuzzer/tlslite-ng/blob/c464c42242f66739176a5d6303b6194380eaf847/tlslite/extensions.py#L2111)):*
<details><summary><i>Previously, tomato42 (Hubert Kario) wrote…</i></summary><blockquote>

`ticket_length` is unused, isn't it?
</blockquote></details>

fixed

File: tlslite/tlsconnection.py [hunk 1]
```diff
@@ -786,6 +787,27 @@ def _clientSendClientHello(self, settings, session, srpUsername,
             extensions.append(RecordSizeLimitExtension().create(
                 settings.record_size_limit))
 
+        # If SessionTicket support is enabled and we have a valid ticket, we
+        # send it in attempt to resume the session, if SessionTicket support
+        # is enabled but we don't have a valid ticket, we send an empty ext
+        # to indicate support fo the feaure
```
Comment (Jakuje) [inline tlslite/tlsconnection.py:793]:
```suggestion
        # to indicate support for the feaure
```

File: tlslite/tlsconnection.py [hunk 2]
```diff
@@ -2619,6 +2671,10 @@ def _serverTLS13Handshake(self, settings, clientHello, cipherSuite,
                         continue
                     match = [match]
 
+                # check if the ticket version matches
+                # but with PSK we don't have a ticket, still have a binder
```
Comment (Jakuje) [inline tlslite/tlsconnection.py:2675]: this comment is hard to read and understand. Can you rephrase?

File: tlslite/tlsconnection.py [hunk 3]
```diff
@@ -4387,17 +4477,47 @@ def _sendFinished(self, masterSecret, cipherSuite=None, nextProto=None,
         self.sock.buffer_writes = False
 
     def _getFinished(self, masterSecret, cipherSuite=None,
-                     expect_next_protocol=False, nextProto=None):
-        #Get and check ChangeCipherSpec
-        for result in self._getMsg(ContentType.change_cipher_spec):
+                    expect_next_protocol=False, nextProto=None):
```
Comment (Jakuje) [inline tlslite/tlsconnection.py:4480]: misindentation?

File: tlslite/extensions.py [hunk 4]
```diff
@@ -2111,6 +2111,53 @@ def __init__(self):
...
+            return self
+        self.ticket = parser.getFixBytes(parser.getRemainingLength())
+
+        return self
+
+    def __repr__(self):
+        """Return human readable representation of the extension."""
+        return "{0}({1}={2!r})".format(self.__class__.__name__,
+                                     "ticket",
+                                     self.ticket)
```
[the remark this answers is not in the document]
Comment (tomato42) [reply tlslite/extensions.py]: fixed

File: tlslite/tlsconnection.py (hunk 1 above)
Comment (tomato42) [reply tlslite/tlsconnection.py]: fixed

File: tlslite/tlsconnection.py (hunk 2 above)
Comment (tomato42) [reply tlslite/tlsconnection.py]: changed

File: tlslite/tlsconnection.py (hunk 3 above)
Comment (tomato42) [reply tlslite/tlsconnection.py]: fixed

**Review by Jakuje — APPROVED** [review]

**Review by tomato42 — APPROVED** [review]
